cpg_pipes/providers/storage.py

```diff
@@ -0,0 +1,138 @@
+"""
+Abstract storage provider.
+"""
+
+import pathlib
+from enum import Enum
+from abc import ABC, abstractmethod
+from typing import Union
+
+from cloudpathlib import CloudPath
+from cloudpathlib.anypath import to_anypath
+
+
+# Path can be either a cloud URL or a local posix file path.
+Path = Union[CloudPath, pathlib.Path]
+
+# Using convenience method from cloudpathlib to parse a path string.
+to_path = to_anypath
+
+
+class Namespace(Enum):
+    """
+    Storage namespace.
+    https://github.com/populationgenomics/team-docs/tree/main/storage_policies#main-vs-test
+    """
+
+    MAIN = 'main'
+    TEST = 'test'
+
+
+class Cloud(Enum):
+    """
+    Cloud storage provider and correponding protocol prefix.
+    """
+
+    GS = 'gs'
+    AZ = 'az'
+
+
+class StorageProvider(ABC):
+    """
+    Abstract class for implementing storage bucket policy.
+    Onty get_bucket() method is required, however other methods
+    are available to override as well.
+    """
+
+    def __init__(self, cloud: Cloud):
+        self.cloud = cloud
+
+    @abstractmethod
+    def get_bucket(
+        self,
+        dataset: str,
+        namespace: Namespace,
+        suffix: str | None = None,
+        version: str | None = None,
+        sample: str = None,
+    ) -> Path:
+        """
+        Bucket to write primary results.
+        @param dataset: dataset/stack name
+        @param namespace: namespace (test or main)
+        @param suffix: (optional) suffix to append to the bucket name
+        @param version: (optional) pipeline version
+        @param sample: (optional) sample name
+        """
+
+    def get_analysis_bucket(
+        self,
+        dataset: str,
+        namespace: Namespace,
+        version: str | None = None,
+        sample: str = None,
+    ) -> Path:
+        """
+        Bucket to write analysis results.
+        """
+        return self.get_bucket(
+            dataset=dataset,
+            namespace=namespace,
+            version=version,
+            sample=sample,
+            suffix='analysis',
+        )
+
+    def get_tmp_bucket(
+        self,
+        dataset: str,
+        namespace: Namespace,
+        version: str | None = None,
+        sample: str = None,
+    ) -> Path:
+        """
+        Bucket to write temporary files.
+        """
+        return self.get_bucket(
+            dataset=dataset,
+            namespace=namespace,
+            version=version,
+            sample=sample,
+            suffix='tmp',
+        )
+
+    def get_web_bucket(
```

Comment:
If we want to support these extra categories, there probably needs to be more documentation that describes when to use them. I wonder if this is too CPG-specific though and we should be aiming for something more abstract, e.g. a `category` that different providers can map however they like. This would be more similar to the current implementation of `output_path` in `cpg-utils`.

Reply:
You could potentially just link to our storage policies even for the abstract class, as that provides a general idea of the categories.